e2e-tests/vm/provision.sh

```diff
@@ -0,0 +1,264 @@
+            exit 1
+    esac
+done
+
+if [ -z "${SSH_PUBLIC_KEY_FILE:-}" ] || \
+   [ -z "${ISSUER_ID:-}" ] || \
+   [ -z "${CLIENT_ID:-}" ]; then
+   echo "Missing required arguments." >&2
+   usage
+   exit 1
+fi
+
+# Validate SSH public key file
+if [ ! -f "${SSH_PUBLIC_KEY_FILE}" ]; then
+    echo "SSH public key file not found: ${SSH_PUBLIC_KEY_FILE}"
+    exit 1
+fi
+
+if [[ "${SSH_PUBLIC_KEY_FILE}" != *.pub ]]; then
+    echo "SSH public key file must have a .pub extension"
+    exit 1
+fi
+
+function cloud_init_finished() {
+    local image=$1
+    sudo guestfish --ro -a "${image}" -i stat /var/lib/cloud/instance/boot-finished &>/dev/null
+}
+
+function force_create_snapshot() {
+    local snapshot_name="$1"
+    if virsh snapshot-list "${VM_NAME}" | grep -q "${snapshot_name}"; then
+        virsh snapshot-delete --domain "${VM_NAME}" --snapshotname "${snapshot_name}"
+    fi
+    virsh snapshot-create-as --domain "${VM_NAME}" "${snapshot_name}" --reuse-external
+}
+
+function wait_for_system_running() {
+    retry --times 10 --delay 3 -- "$SSH" "systemctl is-system-running --wait"
+}
+
+function install_brokers() {
+    local channel="$1"
+    local base_snapshot="authd-${channel}-installed"
+    local broker
+
+    for broker in "${BROKERS[@]}"; do
+        local broker_config="${broker#authd-}.conf"
+
+        # Install broker, configure and restart services
+        $SSH "sudo snap install ${broker} --channel=${channel} && \
+              sudo mkdir -p /etc/authd/brokers.d && \
+              sudo cp /snap/${broker}/current/conf/authd/${broker_config} /etc/authd/brokers.d/ && \
+              sudo sed -i -e 's|<ISSUER_ID>|'${ISSUER_ID}'|g' \
+                          -e 's|<CLIENT_ID>|'${CLIENT_ID}'|g' \
+                          /var/snap/${broker}/current/broker.conf && \
+              sudo systemctl restart authd.service && \
+              sudo snap restart ${broker}"
+
+        # Reboot VM and wait until it's back
+        virsh reboot "${VM_NAME}"
+        wait_for_system_running
+
+        # Snapshot this broker installation
+        force_create_snapshot "${broker}-${channel}-configured"
+
+        # If not the last broker, revert to the base authd snapshot for this channel
+        if [ "${broker}" != "${BROKERS[-1]}" ]; then
+            virsh snapshot-revert "${VM_NAME}" --snapshotname "${base_snapshot}"
+        fi
+    done
+}
+
+# Print executed commands to ease debugging
+set -x
+
+# Install required packages
+sudo apt-get -y install \
+    bsdutils \
+    clang \
+    cloud-image-utils \
+    ffmpeg \
+    gir1.2-webkit2-4.1 \
+    libcairo2-dev \
+    libgirepository-2.0-dev \
+    libvirt0 \
+    libvirt-clients \
+    libvirt-clients-qemu \
+    libvirt-daemon \
+    libvirt-daemon-driver-qemu \
+    libvirt-daemon-system \
+    libxkbcommon-dev \
+    qemu-kvm \
+    qemu-system-x86 \
+    qemu-utils \
+    python3-cairo \
+    python3-gi \
+    python3-tk \
+    socat \
+    sshpass \
+    wget \
+    xvfb
```

Comment:
I would use:

```bash
dependencies=(
    bsdutils
    clang
    cloud-image-utils
    ffmpeg
    gir1.2-webkit2-4.1
    libcairo2-dev
    libgirepository-2.0-dev
    libvirt0
    libvirt-clients
    libvirt-clients-qemu
    libvirt-daemon
    libvirt-daemon-driver-qemu
    libvirt-daemon-system
    libxkbcommon-dev
    qemu-kvm
    qemu-system-x86
    qemu-utils
    python3-cairo
    python3-gi
    python3-tk
    socat
    sshpass
    wget
    xvfb
)

sudo apt-get install -y "${dependencies[@]}"

Reply:
BTW as per some previous discussion, some things like `python3-tk` are not needed at tests runtime, or `lib*-dev`, so can mark them as `auto` and then auto-remove them

Reply:
Well, I understand but we do already have bashism everywhere, so I'd rather explicitly target that (and we can do it with shell check).

I don't expect bash to go anywhere in the next century at least, so if we can use cleaner code it's better IMHO.

And leading backslashes are a thing that is always easy to forget :)